Implement release keybinds and modifier-only binds

flowerysong opened the pull request:
This builds on the work @Glitch752 did in #1923, but takes it in a slightly different direction.

Instead of tying bind validity to (notional, because `Mod` was never actually suppressed) key suppression status and whether any bind has already triggered, we track the most recently pressed key, make it the only one that can trigger a bind on release, and invalidate it when any key is released or any key or mouse button is pressed. This should prevent these binds from unexpectedly triggering when a user is trying to input something else, while allowing intentional binding of the same key to actions on press and release (e.g. for push to talk.)

I also made the implementation of modifier-only binds more generic. Non-`Mod` modifier binds are still configured with keysyms for the trigger (so `Ctrl release=true { toggle-overview; }` won't work but `Control_L release=true { toggle-overview; }` will.)

It seems to function as intended.

Closes #1923, fixes #1813, closes #1148, closes #605.

flowerysong committed:
Allow both press and release binds for the same key

flowerysong committed:
Implement release binds and modifier-only binds

We track the most recently pressed key, make it the only one that can
trigger a bind on release, and invalidate it when any key is released
or any key or mouse button is pressed. This should prevent these binds
from unexpectedly triggering when a user is trying to input something
else.

Modifiers already kind of worked (but were mostly useless without
release binds), we mainly just have to make sure we're passing them
through instead of intercepting them.

Co-authored-by: Glitch752 <xxGlitch752xx@gmail.com>

flowerysong committed:
Allow Mod as a bind trigger

Mod is special so it requires special handling. It would also be
possible to implement syntactic sugar for other modifiers as triggers
but I don't think the current approach using keysyms is bad UX, and it
would complicate detection of duplicate binds.

Co-authored-by: Glitch752 <xxGlitch752xx@gmail.com>

Glitch752 commented:
I'm not great at code review, but I added a few of my thoughts. Otherwise, this looks good to me and seems like a more robust implementation strategy.

Glitch752 commented:
I would add: something to think about is that certain release binds may benefit from overlapping. For example, I would prefer if I could use `Mod release=true` while a push-to-talk key is held. I'm not sure how this implementation could be extended to handle that.

Glitch752 reviewed (commented):
oops! So sorry haha, not sure how I missed that

Glitch752 commented:
> I don't think that's something you should want

This is how I use push-to-talk on other OSes; sometimes I hold down the key and, while talking, press mod. I agree that this is somewhat unintuitive anyway though.

> In general I think making the behaviour predictable and easy for users to reason about is more important than trying to cover absolutely every way people might want it to work

100% agree with this. There's always the option to add more complexity later if it's needed (and in this case it probably isn't anyway). I just wanted to add that in case you hadn't considered it.

flowerysong commented:
> There's always the option to add more complexity later if it's needed (and in this case it probably isn't anyway). I just wanted to add that in case you hadn't considered it.

You need two things to achieve this: a second type of release bind that always triggers instead of caring about validity, and (if you want to support using modifier keys for this) a way to mask certain modifier keys after their bind triggers so they're treated as though they aren't pressed. I don't think the implementation of either part would be difficult.

flowerysong commented on src/input/mod.rs:
```diff
@@ -3999,76 +4079,124 @@ fn should_intercept_key(
 
+/// Finds the specified keybind. Note that release binds can be
+/// returned even when pressed=true because we still need to inhibit
+/// them.
```
Another way to look at it: it always returns a match, it just uses `pressed` as a hint as to which match is preferred when there are multiple. The better way to do this is to return a list, but that would be a more involved change.

flowerysong commented on src/input/mod.rs:
```diff
@@ -569,11 +604,13 @@ impl State {
                 self.backend.change_vt(vt);
                 // Changing VT may not deliver the key releases, so clear the state.
                 self.niri.suppressed_keys.clear();
+                self.niri.last_key = None;
```
If it were a complicated process or something replicated a half-dozen places a refactor might be worthwhile, but with only two places this is done and only two steps I think adding an abstraction would make the code harder to follow for no real benefit.

Glitch752 commented:
Alright, that all makes sense; this looks good to me. I'll switch to this branch on my computer to get another tester, but I tried a lot of release bind combinations and all the behavior feels intuitive/correct.

flowerysong committed:
Add special release binds that always trigger

The main use case (apart from modifier-only binds) the community has
presented for release binds is push-to-talk, which is less viable when
you can trigger the unmute but then accidentally not trigger the mute
because of other input you performed.

hbeberman commented:
I've been using this patch for a few days as a global PTT mute/unmute and it's been working great for me.
```kdl
    KP_Insert { spawn-sh "wpctl set-mute @DEFAULT_AUDIO_SOURCE@ 0"; }
    KP_Insert release=true { spawn-sh "wpctl set-mute @DEFAULT_AUDIO_SOURCE@ 1"; }
```

flowerysong commented:
That line number doesn't make any sense; are you sure you pulled down the latest version of this PR?

It would help if you provided the full context diff, so that we can see the lines surrounding your change.

JohnConnorNPC commented:
> That line number doesn't make any sense; are you sure you pulled down the latest version of this PR?
> 
> It would help if you provided the full context diff, so that we can see the lines surrounding your change.

I hope I got this right.   (It did solve my issue and has not shown any other issues yet)

     @@ -2851,6 +2851,9 @@ impl State {
                  let should_handle =
                      should_handle_in_overview || self.niri.mods_with_wheel_binds.contains(&modifiers);
                  if should_handle {
     +                // Invalidate any pending release trigger, similar to mouse button presses
     +                self.niri.valid_release_trigger = None;
     +
                      let horizontal = horizontal_amount_v120.unwrap_or(0.);
                      let ticks = self.niri.horizontal_wheel_tracker.accumulate(horizontal);
                      if ticks != 0 {

YaLTeR commented:
I haven't looked at this PR in detail yet but the problem posed in the last commit piqued my interest. The special flag solution seems error-prone: the user needs to both know about and remember to set the flag, or risk a potentially privacy-sensitive release bind not triggering.

I have this alternative idea to solve this problem what if press+release binds were defined with this special syntax?

```kdl
binds {
    // Push to talk.
    Ctrl+Alt {
        press { spawn "unmute.sh"; }
        release { spawn "mute.sh"; }
    }
}
```

When handling this kind of bind, if niri runs the `press` action, then it will guarantee running the `release` action, regardless of whether any intermediate keys were pressed.

I would go even further and make this the only way to define release binds, instead of the `release=true` flag, so e.g. you would have:

```kdl
binds {
    // Toggle the Overview on Mod by itself.
    Super {
        release { toggle-overview; }
    }
}
```

This would be equivalent to a sole `release=True` bind under this PR.

By restricting release binds to this syntax, we can also avoid other awkwardness:
- No need for special handling of press/release duplicate key detection. You can only declare press and release on the same key together using this syntax.
- Press and release always stay together in the config, no way for them to get lost.
- Properties apply in an obvious way: `repeat` only to the press bind, `allow-when-locked` and `hotkey-overlay-title` to the combined press/release bind. No chance to mess up by setting `allow-when-locked` only on press and not release for example.

What do you think? Does this still cover all release bind usecases?

miku4k commented:
would that make `press` required for non-release binds?
```kdl
binds {
    Mod+Space {
        // spawn "rofi";
        press { spawn "rofi"; }
    }
}
```

YaLTeR commented:
No. We don't have any actions named press and release so there shouldn't be any conflict

z411 commented:
Thanks for this, I've been using this patch to open overview with my Mod (set to Alt) key no problem:

```
Mod repeat=false release=true { toggle-overview; ]
```

That said, I found an issue when trying to bind it to a specific modifier key, in this case Super:

```
 397 │     Super repeat=false release=true { toggle-overview; }
     ·     ──┬──
     ·       ╰── invalid value
```

I tried Super+Super_L as mentioned by #605 but it does nothing.

flowerysong commented:
It's just `Super_L`.  Since the key has been released the modifier is no longer in effect. See the documentation added in the PR: https://github.com/YaLTeR/niri/pull/2456/files#diff-850fc2afe64e0bb69b9677090803b3571f4889fb3f8c7faa328606f472a28e68R98-R105

`Mod` has special handling so that the two binds can be the same, but trying to do that for the normal modifiers seems very fragile to me.
